File: go/singleton/notification-hub/handler.go
package function

import (
	"fmt"
	"io"
	"net/http"
	"strings"
	"sync"
)

var (
	subscribersMu sync.Mutex
	subscribers   = make(map[chan string]struct{})
)

func Handle(w http.ResponseWriter, r *http.Request) {
	switch r.Method {
	case http.MethodGet:
		subscribe(w, r)
	case http.MethodPost:
		publish(w, r)
	default:
		w.Header().Set("Allow", "GET, POST")
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
	}
}

func subscribe(w http.ResponseWriter, r *http.Request) {
	flusher, ok := w.(http.Flusher)
	if !ok {
		http.Error(
			w,
			"streaming is not supported",
			http.StatusInternalServerError,
		)
		return
	}

	messages := make(chan string, 1)
	subscribersMu.Lock()
	subscribers[messages] = struct{}{}
	subscribersMu.Unlock()

	defer func() {
		subscribersMu.Lock()
		delete(subscribers, messages)
		subscribersMu.Unlock()
	}()

	w.Header().Set("Content-Type", "text/event-stream")
	w.Header().Set("Cache-Control", "no-cache")
	fmt.Fprint(w, ": connected\n\n")
	flusher.Flush()

	for {
		select {
		case message := <-messages:
			fmt.Fprintf(w, "data: %s\n\n", message)
			flusher.Flush()
		case <-r.Context().Done():
			return
		}
	}
}

func publish(w http.ResponseWriter, r *http.Request) {
	defer r.Body.Close()

	body, err := io.ReadAll(r.Body)
	if err != nil {
		http.Error(w, "unable to read request body", http.StatusBadRequest)
		return
	}

	message := strings.TrimSpace(string(body))
	if message == "" {
		http.Error(w, "notification must not be empty", http.StatusBadRequest)
		return
	}

	message = strings.NewReplacer("\r", " ", "\n", " ").Replace(message)
	delivered := broadcast(message)

	w.Header().Set("Content-Type", "application/json")
	fmt.Fprintf(w, "{\"delivered\":%d}\n", delivered)
}

func broadcast(message string) int {
	subscribersMu.Lock()
	defer subscribersMu.Unlock()

	delivered := 0
	for subscriber := range subscribers {
		select {
		case subscriber <- message:
			delivered++
		default:
			// Do not let a slow subscriber block every other client.
		}
	}

	return delivered
}

File: python/singleton/notification-hub/handler.py
import json
import queue
import threading

from flask import Response, request


subscribers = set()
subscribers_lock = threading.Lock()
HEARTBEAT_INTERVAL = 15


def handle(req):
    if request.method == "GET":
        return subscribe()
    if request.method == "POST":
        return publish(req)

    return "method not allowed", 405, {"Allow": "GET, POST"}


def subscribe():
    messages = queue.Queue(maxsize=1)
    with subscribers_lock:
        subscribers.add(messages)

    def stream():
        try:
            yield ": connected\n\n"
            while True:
                try:
                    message = messages.get(timeout=HEARTBEAT_INTERVAL)
                    yield f"data: {message}\n\n"
                except queue.Empty:
                    # Periodic writes let the server detect idle disconnects.
                    yield ": keep-alive\n\n"
        finally:
            with subscribers_lock:
                subscribers.discard(messages)

    return Response(
        stream(),
        mimetype="text/event-stream",
        headers={"Cache-Control": "no-cache"},
    )


def publish(req):
    body = req.decode() if isinstance(req, bytes) else str(req)
    message = " ".join(body.splitlines()).strip()
    if not message:
        return "notification must not be empty", 400

    delivered = broadcast(message)
    return (
        json.dumps({"delivered": delivered}) + "\n",
        200,
        {"Content-Type": "application/json"},
    )


def broadcast(message):
    delivered = 0
    with subscribers_lock:
        for subscriber in subscribers:
            try:
                subscriber.put_nowait(message)
                delivered += 1
            except queue.Full:
                # Do not let a slow subscriber block every other client.
                pass

    return delivered
